diff -Nuarp ruby-2.3.8.a/ext/tk/extconf.rb ruby-2.3.8.b/ext/tk/extconf.rb
--- ruby-2.3.8.a/ext/tk/extconf.rb	2015-12-16 00:31:54.000000000 -0500
+++ ruby-2.3.8.b/ext/tk/extconf.rb	2020-04-04 03:13:35.698648500 -0400
@@ -116,7 +116,7 @@ end
 # fuctions
 ##############################################################
 def is_win32?
-  /mswin|mingw|cygwin|bccwin/ =~ RUBY_PLATFORM
+  /mswin|mingw|bccwin/ =~ RUBY_PLATFORM
 end
 
 def is_macosx?
@@ -554,6 +554,7 @@ def get_ext_list()
   exts = [CONFIG['DLEXT']]
   exts.concat %w(dll) if is_win32?
   exts.concat %w(bundle dylib) if is_macosx?
+  exts.concat %w(dll.a) if /cygwin/ =~ RUBY_PLATFORM
 
   if TkLib_Config["tcltk-stubs"] || TkLib_Config['enable-shared'] == false
     exts.unshift "lib" if is_win32?
@@ -614,7 +615,7 @@ def libcheck_for_tclConfig(tcldir, tkdir
         #puts "check #{file} #{$1} #{tclfunc} #{tcldir}"
         #find_library($1, tclfunc, tcldir)
         if (tclconf && tclconf["TCL_SHARED_BUILD"] == "0") ||
-            (ext != CONFIG['DLEXT'] && ext == CONFIG['LIBEXT']) || ext == "a"
+            (!tclconf && ((ext != CONFIG['DLEXT'] && ext == CONFIG['LIBEXT']) || ext == "a"))
           # static link
           tcllibs = $libs + " -DSTATIC_BUILD " + file.quote
 
@@ -657,7 +658,7 @@ def libcheck_for_tclConfig(tcldir, tkdir
         #puts "check #{file} #{$1} #{tkfunc} #{tkdir}"
         # find_library($1, tkfunc, tkdir)
         if (tkconf && tkconf["TCL_SHARED_BUILD"] == "0") ||
-            (ext != CONFIG['DLEXT'] && ext == CONFIG['LIBEXT']) || ext == "a"
+            (!tkconf && ((ext != CONFIG['DLEXT'] && ext == CONFIG['LIBEXT']) || ext == "a"))
           # static link
           tklibs = " -DSTATIC_BUILD " + file.quote
 
@@ -1823,7 +1824,6 @@ unless is_win32?
   print("\n") # progress
 end
 $CPPFLAGS ||= ""
-$CPPFLAGS += ' -D_WIN32' if /cygwin/ =~ RUBY_PLATFORM
 
 # Does ruby have nativethread ?
 TkLib_Config["ruby_with_thread"] =
diff -Nuarp ruby-2.3.8.a/ext/tk/stubs.c ruby-2.3.8.b/ext/tk/stubs.c
--- ruby-2.3.8.a/ext/tk/stubs.c	2014-09-30 01:25:32.000000000 -0400
+++ ruby-2.3.8.b/ext/tk/stubs.c	2020-04-04 03:14:18.016875400 -0400
@@ -71,7 +71,7 @@ _nativethread_consistency_check(ip)
 
 #if defined USE_TCL_STUBS && defined USE_TK_STUBS
 
-#if defined _WIN32 || defined __CYGWIN__
+#if defined _WIN32
 #  ifdef HAVE_RUBY_RUBY_H
 #    include "ruby/util.h"
 #  else
@@ -100,6 +100,10 @@ _nativethread_consistency_check(ip)
 #  undef DLEXT
 #  define DLEXT ".dylib"
 # endif
+# if defined(__CYGWIN__)
+#  undef DLEXT
+#  define DLEXT ".dll"
+# endif
 #endif
 
 static DL_HANDLE tcl_dll = (DL_HANDLE)0;
