diff -Nuarp ruby-3.3.11.a/abrt.c ruby-3.3.11.b/abrt.c
--- ruby-3.3.11.a/abrt.c	1969-12-31 19:00:00.000000000 -0500
+++ ruby-3.3.11.b/abrt.c	2026-03-29 01:56:34.169233588 -0400
@@ -0,0 +1,12 @@
+#include "internal.h"
+
+void
+Init_abrt(void)
+{
+  rb_eval_string(
+    "  begin\n"
+    "    require 'abrt'\n"
+    "  rescue LoadError\n"
+    "  end\n"
+  );
+}
diff -Nuarp ruby-3.3.11.a/common.mk ruby-3.3.11.b/common.mk
--- ruby-3.3.11.a/common.mk	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/common.mk	2026-03-29 01:56:38.906569723 -0400
@@ -111,7 +111,8 @@ PRISM_FILES = prism/api_node.$(OBJEXT) \
 		prism/prism.$(OBJEXT) \
 		prism_init.$(OBJEXT)
 
-COMMONOBJS    = array.$(OBJEXT) \
+COMMONOBJS    = abrt.$(OBJEXT) \
+                array.$(OBJEXT) \
 		ast.$(OBJEXT) \
 		bignum.$(OBJEXT) \
 		class.$(OBJEXT) \
@@ -1606,8 +1607,6 @@ yes-test-syntax-suggest: $(PREPARE_SYNTA
 	$(ACTIONS_ENDGROUP)
 no-test-syntax-suggest:
 
-check: $(DOT_WAIT) $(PREPARE_SYNTAX_SUGGEST) test-syntax-suggest
-
 test-bundler-precheck: $(TEST_RUNNABLE)-test-bundler-precheck
 no-test-bundler-precheck:
 yes-test-bundler-precheck: main $(arch)-fake.rb
diff -Nuarp ruby-3.3.11.a/configure.ac ruby-3.3.11.b/configure.ac
--- ruby-3.3.11.a/configure.ac	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/configure.ac	2026-03-29 01:56:43.570209026 -0400
@@ -136,7 +136,7 @@ RUBY_BASE_NAME=`echo ruby | sed "$progra
 RUBYW_BASE_NAME=`echo rubyw | sed "$program_transform_name"`
 AC_SUBST(RUBY_BASE_NAME)
 AC_SUBST(RUBYW_BASE_NAME)
-AC_SUBST(RUBY_VERSION_NAME, '${RUBY_BASE_NAME}-${ruby_version}')
+AC_SUBST(RUBY_VERSION_NAME, '${RUBY_BASE_NAME}-${ruby_version_dir_name}')
 
 dnl checks for alternative programs
 AC_CANONICAL_BUILD
@@ -354,7 +354,7 @@ test -z "$warnflags" ||
 AS_IF([test -z "${CFLAGS+set}"], [
     cflags=`echo " $cflags " | sed "$cflagspat;s/^ *//;s/ *$//"`
     orig_cflags="$cflags"
-    cflags="$cflags "'${optflags} ${debugflags} ${warnflags}'
+    cflags='${hardenflags} '"$cflags "'${optflags} ${debugflags} ${warnflags}'
 ])
 dnl AS_IF([test -z "${CXXFLAGS+set}"], [
 dnl     cxxflags=`echo " $cxxflags " | sed "$cflagspat;s/^ *//;s/ *$//"`
@@ -802,7 +802,7 @@ AS_IF([test "$GCC" = yes], [
 		  [fortify_source=$enableval])
     AS_IF([test "x$fortify_source" != xno], [
         RUBY_TRY_CFLAGS([$optflags -D_FORTIFY_SOURCE=2],
-                        [RUBY_APPEND_OPTION(XCFLAGS, -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2)], [],
+                        [RUBY_PREPEND_OPTION(hardenflags, -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2)], [],
                         [@%:@include <stdio.h>])
     ])
 
@@ -823,20 +823,24 @@ AS_IF([test "$GCC" = yes], [
     AC_MSG_CHECKING([for -fstack-protector])
     AC_MSG_RESULT(["$stack_protector"])
     AS_CASE(["$stack_protector"], [-*], [
-	RUBY_APPEND_OPTION(XCFLAGS, $stack_protector)
-	RUBY_APPEND_OPTION(XLDFLAGS, $stack_protector)
-	RUBY_APPEND_OPTION(LDFLAGS, $stack_protector)
+        RUBY_PREPEND_OPTION(hardenflags, $stack_protector)
+        RUBY_APPEND_OPTION(XLDFLAGS, $stack_protector)
+        RUBY_APPEND_OPTION(LDFLAGS, $stack_protector)
     ])
 
     # aarch64 branch protection
     AS_CASE(["$target_cpu"], [aarch64], [
 	AS_FOR(option, opt, [-mbranch-protection=pac-ret -msign-return-address=all], [
-            RUBY_TRY_CFLAGS(option, [branch_protection=yes], [branch_protection=no])
+            # Try these flags in the _prepended_ position - i.e. we want to try building a program
+            # with CFLAGS="-mbranch-protection=pac-ret $CFLAGS". If the builder has provided different
+            # branch protection flags in CFLAGS, we don't want to overwrite those. We just want to
+            # find some branch protection flags which work if none were provided.
+            RUBY_TRY_CFLAGS_PREPEND(option, [branch_protection=yes], [branch_protection=no])
             AS_IF([test "x$branch_protection" = xyes], [
-                # C compiler and assembler must be consistent for -mbranch-protection
-                # since they both check `__ARM_FEATURE_PAC_DEFAULT` definition.
-                RUBY_APPEND_OPTION(XCFLAGS, option)
-                RUBY_APPEND_OPTION(ASFLAGS, option)
+                # _prepend_ the options to CFLAGS, so that user-provided flags will overwrite them.
+                # These CFLAGS are used during the configure script to compile further test programs;
+                # however, $harden_flags is prepended separately to CFLAGS at the end of the script.
+                RUBY_PREPEND_OPTION(hardenflags, $opt)
                 break
             ])
         ])
@@ -985,6 +989,59 @@ test -z "${ac_env_CFLAGS_set}" -a -n "${
 test -z "${ac_env_CXXFLAGS_set}" -a -n "${cxxflags+set}" && eval CXXFLAGS="\"$cxxflags $ARCH_FLAG\""
 }
 
+# The lines above expand out the $cflags/$optflags/$debugflags/$hardenflags variables into the
+# CFLAGS variable. So, at this point, we have a $CFLAGS var with the actual compiler flags we're
+# going to use.
+# That means this is the right time to check what branch protection flags are going to be in use
+# and define appropriate macros for use in Context.S based on this
+AS_CASE(["$target_cpu"], [aarch64], [
+    AC_CACHE_CHECK([whether __ARM_FEATURE_BTI_DEFAULT is defined],
+        rb_cv_aarch64_bti_enabled,
+        AC_COMPILE_IFELSE(
+            [AC_LANG_PROGRAM([[
+                @%:@ifndef __ARM_FEATURE_BTI_DEFAULT
+                @%:@error "__ARM_FEATURE_BTI_DEFAULT not defined"
+                @%:@endif
+            ]])],
+        [rb_cv_aarch64_bti_enabled=yes],
+        [rb_cv_aarch64_bti_enabled=no])
+    )
+    AS_IF([test "$rb_cv_aarch64_bti_enabled" = yes],
+          AC_DEFINE(RUBY_AARCH64_BTI_ENABLED, 1))
+    AC_CACHE_CHECK([whether __ARM_FEATURE_PAC_DEFAULT is defined],
+        rb_cv_aarch64_pac_enabled,
+        AC_COMPILE_IFELSE(
+            [AC_LANG_PROGRAM([[
+                @%:@ifndef __ARM_FEATURE_PAC_DEFAULT
+                @%:@error "__ARM_FEATURE_PAC_DEFAULT not defined"
+                @%:@endif
+            ]])],
+        [rb_cv_aarch64_pac_enabled=yes],
+        [rb_cv_aarch64_pac_enabled=no])
+    )
+    AS_IF([test "$rb_cv_aarch64_pac_enabled" = yes],
+          AC_DEFINE(RUBY_AARCH64_PAC_ENABLED, 1))
+    # Context.S will only ever sign its return address with the A-key; it doesn't support
+    # the B-key at the moment.
+    AS_IF([test "$rb_cv_aarch64_pac_enabled" = yes], [
+        AC_CACHE_CHECK([whether __ARM_FEATURE_PAC_DEFAULT specifies the b-key bit 0x02],
+            rb_cv_aarch64_pac_b_key,
+            AC_COMPILE_IFELSE(
+                [AC_LANG_PROGRAM([[
+                    @%:@ifdef __ARM_FEATURE_PAC_DEFAULT
+                    @%:@if __ARM_FEATURE_PAC_DEFAULT & 0x02
+                    @%:@error "__ARM_FEATURE_PAC_DEFAULT specifies B key"
+                    @%:@endif
+                    @%:@endif
+                ]])],
+            [rb_cv_aarch64_pac_b_key=no],
+            [rb_cv_aarch64_pac_b_key=yes])
+        )
+        AS_IF([test "$rb_cv_aarch64_pac_b_key" = yes],
+            AC_MSG_ERROR(-mbranch-protection flag specified b-key but Ruby's Context.S does not support this yet.))
+    ])
+])
+
 AC_CACHE_CHECK([whether compiler has statement and declarations in expressions],
   rb_cv_have_stmt_and_decl_in_expr,
   [AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[]],[[ __extension__ ({ int a = 0; a; }); ]])],
@@ -3480,6 +3537,11 @@ AS_IF([test ${multiarch+set}], [
 ])
 
 archlibdir='${libdir}/${arch}'
+AC_ARG_WITH(archlibdir,
+	    AS_HELP_STRING([--with-archlibdir=DIR],
+			   [prefix for libruby [[LIBDIR/ARCH]]]),
+	    [archlibdir="$withval"])
+
 sitearchlibdir='${libdir}/${sitearch}'
 archincludedir='${includedir}/${arch}'
 sitearchincludedir='${includedir}/${sitearch}'
@@ -4209,12 +4271,13 @@ AS_IF([test "${ARCH_FLAG}"], [
 rb_cv_warnflags=`echo "$rb_cv_warnflags" | sed 's/^ *//;s/ *$//'`
 warnflags="$rb_cv_warnflags"
 AC_SUBST(cppflags)dnl
-AC_SUBST(cflags, ["${orig_cflags:+$orig_cflags }"'${optflags} ${debugflags} ${warnflags}'])dnl
+AC_SUBST(cflags, ['${hardenflags} '"${orig_cflags:+$orig_cflags }"' ${optflags} ${debugflags} ${warnflags}'])dnl
 AC_SUBST(cxxflags)dnl
 AC_SUBST(optflags)dnl
 AC_SUBST(debugflags)dnl
 AC_SUBST(warnflags)dnl
 AC_SUBST(strict_warnflags)dnl
+AC_SUBST(hardenflags)dnl
 AC_SUBST(XCFLAGS)dnl
 AC_SUBST(XLDFLAGS)dnl
 AC_SUBST(EXTLDFLAGS)dnl
@@ -4270,9 +4333,6 @@ AS_CASE(["$target_os"],
     rubyw_install_name='$(RUBYW_INSTALL_NAME)'
     ])
 
-rubylibdir='${rubylibprefix}/${ruby_version}'
-rubyarchdir=${multiarch+'${rubyarchprefix}/${ruby_version}'}${multiarch-'${rubylibdir}/${arch}'}
-
 rubyarchprefix=${multiarch+'${archlibdir}/${RUBY_BASE_NAME}'}${multiarch-'${rubylibprefix}/${arch}'}
 AC_ARG_WITH(rubyarchprefix,
 	    AS_HELP_STRING([--with-rubyarchprefix=DIR],
@@ -4295,57 +4355,68 @@ AC_ARG_WITH(ridir,
 AC_SUBST(ridir)
 AC_SUBST(RI_BASE_NAME)
 
-AC_ARG_WITH(ruby-version,
-	    AS_HELP_STRING([--with-ruby-version=STR], [ruby version string for version specific directories [[full]] (full|minor|STR)]),
-            [ruby_version=$withval],
-            [ruby_version=full])
 unset RUBY_LIB_VERSION
-unset RUBY_LIB_VERSION_STYLE
-AS_CASE(["$ruby_version"],
-  [full],  [RUBY_LIB_VERSION_STYLE='3	/* full */'],
-  [minor], [RUBY_LIB_VERSION_STYLE='2	/* minor */'])
-AS_IF([test ${RUBY_LIB_VERSION_STYLE+set}], [
-    {
-    echo "#define RUBY_LIB_VERSION_STYLE $RUBY_LIB_VERSION_STYLE"
-    echo '@%:@include "confdefs.h"'
-    echo '#define STRINGIZE(x) x'
-    test -f revision.h -o -f "${srcdir}/revision.h" || echo '#define RUBY_REVISION 0'
-    echo '#include "version.h"'
-    echo 'ruby_version=RUBY_LIB_VERSION'
-    } > conftest.c
-    ruby_version="`$CPP -I. -I"${srcdir}" -I"${srcdir}/include" conftest.c | sed '/^ruby_version=/!d;s/ //g'`"
-    eval $ruby_version
-], [test -z "${ruby_version}"], [
-    AC_MSG_ERROR([No ruby version, No place for bundled libraries])
-], [
-    RUBY_LIB_VERSION="${ruby_version}"
-])
+RUBY_LIB_VERSION_STYLE='3	/* full */'
+{
+echo "#define RUBY_LIB_VERSION_STYLE $RUBY_LIB_VERSION_STYLE"
+echo '@%:@include "confdefs.h"'
+echo '#define STRINGIZE(x) x'
+test -f revision.h -o -f "${srcdir}/revision.h" || echo '#define RUBY_REVISION 0'
+echo '#include "version.h"'
+echo 'ruby_version=RUBY_LIB_VERSION'
+} > conftest.c
+ruby_version="`$CPP -I. -I"${srcdir}" -I"${srcdir}/include" conftest.c | sed '/^ruby_version=/!d;s/ //g'`"
+eval $ruby_version
+
+RUBY_LIB_VERSION="${ruby_version}"
+
 AC_SUBST(RUBY_LIB_VERSION_STYLE)
 AC_SUBST(RUBY_LIB_VERSION)
 
+AC_ARG_WITH(ruby-version,
+	    AS_HELP_STRING([--with-ruby-version=STR], [ruby version string for version specific directories [[full]] (full|STR)]),
+            [ruby_version_dir_name=$withval],
+            [ruby_version_dir_name=full])
+AS_CASE(["$ruby_version_dir_name"],
+  [full], [ruby_version_dir_name='${ruby_version}'])
+
+ruby_version_dir=/'${ruby_version_dir_name}'
+
+if test -z "${ruby_version_dir_name}"; then
+    unset ruby_version_dir
+    AC_DEFINE(RUBY_LIB_VERSION_BLANK, 1)
+fi
+
+rubylibdir='${rubylibprefix}'${ruby_version_dir}
+rubyarchdir=${multiarch+'${rubyarchprefix}'${ruby_version_dir}}${multiarch-'${rubylibdir}/${arch}'}
+
 AC_ARG_WITH(sitedir,
 	    AS_HELP_STRING([--with-sitedir=DIR], [site libraries in DIR [[RUBY_LIB_PREFIX/site_ruby]], "no" to disable site directory]),
             [sitedir=$withval],
             [sitedir='${rubylibprefix}/site_ruby'])
-sitelibdir='${sitedir}/${ruby_version}'
+sitelibdir='${sitedir}'${ruby_version_dir}
 
 AC_ARG_WITH(sitearchdir,
 	    AS_HELP_STRING([--with-sitearchdir=DIR],
 			   [architecture dependent site libraries in DIR [[SITEDIR/SITEARCH]], "no" to disable site directory]),
             [sitearchdir=$withval],
-            [sitearchdir=${multiarch+'${rubysitearchprefix}/site_ruby/${ruby_version}'}${multiarch-'${sitelibdir}/${sitearch}'}])
+            [sitearchdir=${multiarch+'${rubysitearchprefix}/site_ruby'${ruby_version_dir}}${multiarch-'${sitelibdir}/${sitearch}'}])
 
 AC_ARG_WITH(vendordir,
 	    AS_HELP_STRING([--with-vendordir=DIR], [vendor libraries in DIR [[RUBY_LIB_PREFIX/vendor_ruby]], "no" to disable vendor directory]),
             [vendordir=$withval],
             [vendordir='${rubylibprefix}/vendor_ruby'])
-vendorlibdir='${vendordir}/${ruby_version}'
+vendorlibdir='${vendordir}'${ruby_version_dir}
 
 AC_ARG_WITH(vendorarchdir,
 	    AS_HELP_STRING([--with-vendorarchdir=DIR],
 			   [architecture dependent vendor libraries in DIR [[VENDORDIR/SITEARCH]], "no" to disable vendor directory]),
             [vendorarchdir=$withval],
-            [vendorarchdir=${multiarch+'${rubysitearchprefix}/vendor_ruby/${ruby_version}'}${multiarch-'${vendorlibdir}/${sitearch}'}])
+            [vendorarchdir=${multiarch+'${rubysitearchprefix}/vendor_ruby'${ruby_version_dir}}${multiarch-'${vendorlibdir}/${sitearch}'}])
+
+AC_ARG_WITH(rubygemsdir,
+           AS_HELP_STRING([--with-rubygemsdir=DIR], [custom rubygems directory]),
+            [rubygemsdir=$withval])
 
 AS_IF([test "${LOAD_RELATIVE+set}"], [
     AC_DEFINE_UNQUOTED(LOAD_RELATIVE, $LOAD_RELATIVE)
@@ -4362,6 +4433,7 @@ AC_SUBST(sitearchincludedir)dnl
 AC_SUBST(arch)dnl
 AC_SUBST(sitearch)dnl
 AC_SUBST(ruby_version)dnl
+AC_SUBST(ruby_version_dir_name)dnl
 AC_SUBST(rubylibdir)dnl
 AC_SUBST(rubyarchdir)dnl
 AC_SUBST(sitedir)dnl
@@ -4370,10 +4442,13 @@ AC_SUBST(sitearchdir)dnl
 AC_SUBST(vendordir)dnl
 AC_SUBST(vendorlibdir)dnl
 AC_SUBST(vendorarchdir)dnl
+AC_SUBST(rubygemsdir)dnl
 
 AC_SUBST(CONFIGURE, "`echo $0 | sed 's|.*/||'`")dnl
 AC_SUBST(configure_args, "`echo "${ac_configure_args}" | sed 's/\\$/$$/g'`")dnl
 
+target_cpu=`echo $target_cpu | sed s/i.86/i386/`
+
 AS_IF([test "${universal_binary-no}" = yes ], [
     arch="universal-${target_os}"
     AS_IF([test "${rb_cv_architecture_available}" = yes], [
@@ -4670,6 +4745,7 @@ config_summary "DLDFLAGS"            "$D
 config_summary "optflags"            "$optflags"
 config_summary "debugflags"          "$debugflags"
 config_summary "warnflags"           "$warnflags"
+config_summary "hardenflags"         "$hardenflags"
 config_summary "strip command"       "$STRIP"
 config_summary "install doc"         "$DOCTARGETS"
 config_summary "YJIT support"        "$YJIT_SUPPORT"
diff -Nuarp ruby-3.3.11.a/coroutine/arm64/Context.S ruby-3.3.11.b/coroutine/arm64/Context.S
--- ruby-3.3.11.a/coroutine/arm64/Context.S	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/coroutine/arm64/Context.S	2026-03-29 01:56:43.570493890 -0400
@@ -5,6 +5,8 @@
 ##  Copyright, 2018, by Samuel Williams.
 ##
 
+#include "ruby/config.h"
+
 #define TOKEN_PASTE(x,y) x##y
 #define PREFIXED_SYMBOL(prefix,name) TOKEN_PASTE(prefix,name)
 
@@ -27,10 +29,10 @@
 .global PREFIXED_SYMBOL(SYMBOL_PREFIX,coroutine_transfer)
 PREFIXED_SYMBOL(SYMBOL_PREFIX,coroutine_transfer):
 
-#if defined(__ARM_FEATURE_PAC_DEFAULT) && (__ARM_FEATURE_PAC_DEFAULT != 0)
+#if defined(RUBY_AARCH64_PAC_ENABLED)
 	# paciasp (it also acts as BTI landing pad, so no need to insert BTI also)
 	hint #25
-#elif defined(__ARM_FEATURE_BTI_DEFAULT) && (__ARM_FEATURE_BTI_DEFAULT != 0)
+#elif defined(RUBY_AARCH64_BTI_ENABLED)
 	# For the the case PAC is not enabled but BTI is.
 	# bti c
 	hint #34
@@ -73,7 +75,7 @@ PREFIXED_SYMBOL(SYMBOL_PREFIX,coroutine_
 	# Pop stack frame
 	add sp, sp, 0xa0
 
-#if defined(__ARM_FEATURE_PAC_DEFAULT) && (__ARM_FEATURE_PAC_DEFAULT != 0)
+#if defined(RUBY_AARCH64_PAC_ENABLED)
 	# autiasp: Authenticate x30 (LR) with SP and key A
 	hint #29
 #endif
@@ -85,18 +87,18 @@ PREFIXED_SYMBOL(SYMBOL_PREFIX,coroutine_
 .section .note.GNU-stack,"",%progbits
 #endif
 
-#if __ARM_FEATURE_BTI_DEFAULT != 0 || __ARM_FEATURE_PAC_DEFAULT != 0
+#if defined(RUBY_AARCH64_BTI_ENABLED) || defined(RUBY_AARCH64_PAC_ENABLED)
 /*  See "ELF for the Arm 64-bit Architecture (AArch64)"
     https://github.com/ARM-software/abi-aa/blob/2023Q3/aaelf64/aaelf64.rst#program-property */
 #  define GNU_PROPERTY_AARCH64_FEATURE_1_BTI (1<<0)
 #  define GNU_PROPERTY_AARCH64_FEATURE_1_PAC (1<<1)
 
-#  if __ARM_FEATURE_BTI_DEFAULT != 0
+#  if defined(RUBY_AARCH64_BTI_ENABLED)
 #    define BTI_FLAG GNU_PROPERTY_AARCH64_FEATURE_1_BTI
 #  else
 #    define BTI_FLAG 0
 #  endif
-#  if __ARM_FEATURE_PAC_DEFAULT != 0
+#  if defined(RUBY_AARCH64_PAC_ENABLED)
 #    define PAC_FLAG GNU_PROPERTY_AARCH64_FEATURE_1_PAC
 #  else
 #    define PAC_FLAG 0
diff -Nuarp ruby-3.3.11.a/ext/openssl/lib/openssl/ssl.rb ruby-3.3.11.b/ext/openssl/lib/openssl/ssl.rb
--- ruby-3.3.11.a/ext/openssl/lib/openssl/ssl.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/ext/openssl/lib/openssl/ssl.rb	2026-03-29 01:56:48.466819496 -0400
@@ -22,7 +22,6 @@ module OpenSSL
   module SSL
     class SSLContext
       DEFAULT_PARAMS = { # :nodoc:
-        :min_version => OpenSSL::SSL::TLS1_VERSION,
         :verify_mode => OpenSSL::SSL::VERIFY_PEER,
         :verify_hostname => true,
         :options => -> {
@@ -55,6 +54,7 @@ ssbzSibBsu/6iGtCOGEoXJf//////////wIBAg==
       if !(OpenSSL::OPENSSL_VERSION.start_with?("OpenSSL") &&
            OpenSSL::OPENSSL_VERSION_NUMBER >= 0x10100000)
         DEFAULT_PARAMS.merge!(
+          min_version: OpenSSL::SSL::TLS1_VERSION,
           ciphers: %w{
             ECDHE-ECDSA-AES128-GCM-SHA256
             ECDHE-RSA-AES128-GCM-SHA256
diff -Nuarp ruby-3.3.11.a/gc.c ruby-3.3.11.b/gc.c
--- ruby-3.3.11.a/gc.c	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/gc.c	2026-03-29 01:56:52.897241278 -0400
@@ -824,6 +824,8 @@ typedef struct rb_objspace {
         unsigned int during_minor_gc : 1;
         unsigned int during_incremental_marking : 1;
         unsigned int measure_gc : 1;
+        unsigned int collect_gc_stats : 1;
+        unsigned int verbose_gc_stats : 1;
     } flags;
 
     rb_event_flag_t hook_events;
@@ -890,11 +892,18 @@ typedef struct rb_objspace {
 
         /* temporary profiling space */
         double gc_sweep_start_time;
+        double gc_mark_start_time;
+
         size_t total_allocated_objects_at_gc_start;
         size_t heap_used_at_gc_start;
 
         /* basic statistics */
         size_t count;
+        double time;
+        size_t total_mallocs;
+        size_t total_malloced_bytes;
+        size_t live_after_last_sweep;
+
         uint64_t marking_time_ns;
         struct timespec marking_start_time;
         uint64_t sweeping_time_ns;
@@ -1153,6 +1162,7 @@ VALUE *ruby_initial_gc_stress_ptr = &rub
 #define heap_pages_deferred_final	objspace->heap_pages.deferred_final
 #define size_pools              objspace->size_pools
 #define during_gc		objspace->flags.during_gc
+#define collect_gc_stats        objspace->flags.collect_gc_stats
 #define finalizing		objspace->atomic_flags.finalizing
 #define finalizer_table 	objspace->finalizer_table
 #define global_list		objspace->global_list
@@ -9387,6 +9397,107 @@ heap_ready_to_gc(rb_objspace_t *objspace
     }
 }
 
+VALUE
+rb_gc_enable_stats(void)
+{
+    rb_objspace_t *objspace = &rb_objspace;
+    int old = collect_gc_stats;
+    collect_gc_stats = 1;
+    return old ? Qtrue : Qfalse;
+}
+
+static VALUE
+gc_enable_stats(rb_execution_context_t *ec, VALUE _)
+{
+    return rb_gc_enable_stats();
+}
+
+VALUE
+rb_gc_disable_stats(void)
+{
+    rb_objspace_t *objspace = &rb_objspace;
+    int old = collect_gc_stats;
+    collect_gc_stats = 0;
+    return old ? Qtrue : Qfalse;
+}
+
+static VALUE
+gc_disable_stats(rb_execution_context_t *ec, VALUE _)
+{
+    return rb_gc_disable_stats();
+}
+
+
+VALUE
+rb_gc_stats_enabled(void)
+{
+    rb_objspace_t *objspace = &rb_objspace;
+    return collect_gc_stats ? Qtrue : Qfalse;
+}
+
+static VALUE
+gc_stats_enabled(rb_execution_context_t *ec, VALUE _)
+{
+    return rb_gc_stats_enabled();
+}
+
+double rb_gc_total_time(void)
+{
+    return rb_objspace.profile.time;
+}
+
+static VALUE
+gc_time(rb_execution_context_t *ec, VALUE _)
+{
+    return DBL2NUM(1000000*rb_objspace.profile.time);
+}
+
+VALUE
+rb_gc_heap_slots(void)
+{
+    rb_objspace_t *objspace = &rb_objspace;
+    return SIZET2NUM(heap_allocated_pages * HEAP_PAGE_OBJ_LIMIT);
+}
+
+static
+VALUE gc_heap_slots(rb_execution_context_t *ec, VALUE _)
+{
+    return rb_gc_heap_slots();
+}
+
+VALUE
+rb_gc_heap_slots_live_after_last_gc(void)
+{
+    rb_objspace_t *objspace = &rb_objspace;
+    return SIZET2NUM(objspace->profile.live_after_last_sweep);
+}
+
+static
+VALUE gc_heap_slots_live_after_last_gc(rb_execution_context_t *ec, VALUE _)
+{
+    return rb_gc_heap_slots_live_after_last_gc();
+}
+
+size_t rb_gc_total_mallocs(void) {
+    return rb_objspace.profile.total_mallocs;
+}
+
+static VALUE
+gc_total_mallocs(rb_execution_context_t *ec, VALUE _)
+{
+    return SIZET2NUM(rb_objspace.profile.total_mallocs);
+}
+
+size_t rb_gc_total_malloced_bytes(void) {
+    return rb_objspace.profile.total_malloced_bytes;
+}
+
+static VALUE
+gc_total_malloced_bytes(rb_execution_context_t *ec, VALUE _)
+{
+    return SIZET2NUM(rb_objspace.profile.total_malloced_bytes);
+}
+
 static int
 ready_to_gc(rb_objspace_t *objspace)
 {
@@ -12433,6 +12544,10 @@ objspace_malloc_increase_body(rb_objspac
 {
     if (new_size > old_size) {
         ATOMIC_SIZE_ADD(malloc_increase, new_size - old_size);
+        if (collect_gc_stats) {
+            ATOMIC_SIZE_ADD(objspace->profile.total_mallocs, 1);
+            ATOMIC_SIZE_ADD(objspace->profile.total_malloced_bytes, new_size - old_size);
+        }
 #if RGENGC_ESTIMATE_OLDMALLOC
         ATOMIC_SIZE_ADD(objspace->rgengc.oldmalloc_increase, new_size - old_size);
 #endif
@@ -13080,6 +13195,12 @@ gc_malloc_allocated_size(VALUE self)
     return UINT2NUM(rb_objspace.malloc_params.allocated_size);
 }
 
+size_t
+rb_gc_malloc_allocated_size(void)
+{
+    return rb_objspace.malloc_params.allocated_size;
+}
+
 /*
  *  call-seq:
  *     GC.malloc_allocations -> Integer
@@ -13094,6 +13215,12 @@ gc_malloc_allocations(VALUE self)
 {
     return UINT2NUM(rb_objspace.malloc_params.allocations);
 }
+
+size_t
+rb_gc_malloc_allocations(void)
+{
+    return rb_objspace.malloc_params.allocations;
+}
 #endif
 
 void
@@ -13266,6 +13393,14 @@ gc_prof_mark_timer_start(rb_objspace_t *
 #if GC_PROFILE_MORE_DETAIL
     if (gc_prof_enabled(objspace)) {
         gc_prof_record(objspace)->gc_mark_time = getrusage_time();
+    } else {
+        if (collect_gc_stats) {
+          objspace->profile.gc_mark_start_time = getrusage_time();
+        }
+    }
+#else
+    if (collect_gc_stats) {
+        objspace->profile.gc_mark_start_time = getrusage_time();
     }
 #endif
 }
@@ -13278,6 +13413,17 @@ gc_prof_mark_timer_stop(rb_objspace_t *o
     if (gc_prof_enabled(objspace)) {
         gc_profile_record *record = gc_prof_record(objspace);
         record->gc_mark_time = elapsed_time_from(record->gc_mark_time);
+        if (collect_gc_stats) {
+            objspace->profile.time += record->gc_mark_time;
+        }
+    } else {
+        if (collect_gc_stats) {
+            objspace->profile.time += elapsed_time_from(objspace->profile.gc_mark_start_time);
+        }
+    }
+#else
+    if (collect_gc_stats) {
+        objspace->profile.time += elapsed_time_from(objspace->profile.gc_mark_start_time);
     }
 #endif
 }
@@ -13289,7 +13435,11 @@ gc_prof_sweep_timer_start(rb_objspace_t
     if (gc_prof_enabled(objspace)) {
         gc_profile_record *record = gc_prof_record(objspace);
 
-        if (record->gc_time > 0 || GC_PROFILE_MORE_DETAIL) {
+        if (record->gc_time > 0 || GC_PROFILE_MORE_DETAIL || collect_gc_stats) {
+            objspace->profile.gc_sweep_start_time = getrusage_time();
+        }
+    } else {
+        if (collect_gc_stats) {
             objspace->profile.gc_sweep_start_time = getrusage_time();
         }
     }
@@ -13308,9 +13458,19 @@ gc_prof_sweep_timer_stop(rb_objspace_t *
             sweep_time = elapsed_time_from(objspace->profile.gc_sweep_start_time);
             /* need to accumulate GC time for lazy sweep after gc() */
             record->gc_time += sweep_time;
+            if (collect_gc_stats) {
+                objspace->profile.time += sweep_time;
+            }
         }
         else if (GC_PROFILE_MORE_DETAIL) {
             sweep_time = elapsed_time_from(objspace->profile.gc_sweep_start_time);
+            if (collect_gc_stats) {
+		objspace->profile.time += sweep_time;
+	    }
+	} else {
+	    if (collect_gc_stats) {
+		objspace->profile.time += elapsed_time_from(objspace->profile.gc_sweep_start_time);
+	    }
         }
 
 #if GC_PROFILE_MORE_DETAIL
@@ -13318,6 +13478,10 @@ gc_prof_sweep_timer_stop(rb_objspace_t *
         if (heap_pages_deferred_final) record->flags |= GPR_FLAG_HAVE_FINALIZE;
 #endif
         if (heap_pages_deferred_final) objspace->profile.latest_gc_info |= GPR_FLAG_HAVE_FINALIZE;
+    } else {
+        if (collect_gc_stats) {
+            objspace->profile.time += elapsed_time_from(objspace->profile.gc_sweep_start_time);
+        }
     }
 }
 
@@ -13336,9 +13500,13 @@ gc_prof_set_malloc_info(rb_objspace_t *o
 static inline void
 gc_prof_set_heap_info(rb_objspace_t *objspace)
 {
+    if (objspace->profile.total_allocated_objects_at_gc_start > total_freed_objects(objspace))
+	objspace->profile.live_after_last_sweep =
+	    objspace->profile.total_allocated_objects_at_gc_start - total_freed_objects(objspace);
+
     if (gc_prof_enabled(objspace)) {
         gc_profile_record *record = gc_prof_record(objspace);
-        size_t live = objspace->profile.total_allocated_objects_at_gc_start - total_freed_objects(objspace);
+        size_t live = objspace->profile.live_after_last_sweep;
         size_t total = objspace->profile.heap_used_at_gc_start * HEAP_PAGE_OBJ_LIMIT;
 
 #if GC_PROFILE_MORE_DETAIL
diff -Nuarp ruby-3.3.11.a/gc.rb ruby-3.3.11.b/gc.rb
--- ruby-3.3.11.a/gc.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/gc.rb	2026-03-29 01:56:52.897433711 -0400
@@ -320,6 +320,92 @@ module GC
       ULL2NUM(rb_objspace.profile.marking_time_ns + rb_objspace.profile.sweeping_time_ns)
     }
   end
+
+  # call-seq:
+  #    GC.enable_stats	  => true or false
+  #
+  # Enables garbage collection statistics, returning <code>true</code> if garbage
+  # collection statistics was already enabled.
+  #
+  #    GC.enable_stats	 #=> false or true
+  #    GC.enable_stats	 #=> true
+  def self.enable_stats
+    Primitive.gc_enable_stats
+  end
+
+  #   call-seq:
+  #    GC.disable_stats	   => true or false
+  #
+  # Disables garbage collection statistics, returning <code>true</code> if garbage
+  # collection statistics was already disabled.
+  #
+  #    GC.disable_stats	  #=> false or true
+  #    GC.disable_stats	  #=> true
+  def self.disable_stats
+    Primitive.gc_disable_stats
+  end
+
+  # call-seq:
+  #    GC.stats_enabled?    => true or false
+  #
+  # Check whether GC stats have been enabled.
+  #
+  #    GC.stats_enabled?   #=> false or true
+  def self.stats_enabled?
+    Primitive.gc_stats_enabled
+  end
+
+  #   call-seq:
+  #    GC.time	  => Integer
+  #
+  # Returns the time spent during garbage collection while GC statistics collection
+  # was enabled (in micro seconds).
+  #
+  #    GC.time	  #=> 20000
+  def self.time
+    Primitive.gc_time
+  end
+
+  #  call-seq:
+  #	GC.heap_slots	=> Integer
+  #
+  #  Returns the number of heap slots available for object allocations.
+  #
+  #	GC.heap_slots	#=> 10000
+  def self.heap_slots
+    Primitive.gc_heap_slots
+  end
+
+  # call-seq:
+  #    GC.heap_slots_live_after_last_gc	   => Integer
+  #
+  # Returns the number of heap slots which were live after the last garbage collection.
+  #
+  #    GC.heap_slots_live_after_last_gc	   #=> 231223
+  def self.heap_slots_live_after_last_gc
+    Primitive.gc_heap_slots_live_after_last_gc
+  end
+
+  #   call-seq:
+  #	 GC.total_mallocs	   => Integer
+  #
+  #   Returns the number malloc calls. Might wrap around.
+  #
+  #	 GC.total_mallocs	   #=> 324234323246
+  def self.total_mallocs
+    Primitive.gc_total_mallocs
+  end
+
+  #   call-seq:
+  #	 GC.total_malloced_bytes	   => Integer
+  #
+  #   Returns the number of bytes allocated. Might wrap around.
+  #
+  #	 GC.total_malloced_bytes	   #=> 354656256432446
+  def self.total_malloced_bytes
+    Primitive.gc_total_malloced_bytes
+  end
+
 end
 
 module ObjectSpace
diff -Nuarp ruby-3.3.11.a/gc.rbinc ruby-3.3.11.b/gc.rbinc
--- ruby-3.3.11.a/gc.rbinc	2026-03-25 20:05:06.000000000 -0400
+++ ruby-3.3.11.b/gc.rbinc	2026-03-29 01:57:02.058761868 -0400
@@ -52,6 +52,14 @@ void Init_builtin_gc(void)
     RB_BUILTIN_FUNCTION(10, _bi0, builtin_inline_class_297, 0),
     RB_BUILTIN_FUNCTION(11, _bi1, builtin_inline_class_309, 0),
     RB_BUILTIN_FUNCTION(12, _bi2, builtin_inline_class_319, 0),
+    RB_BUILTIN_FUNCTION(13, gc_enable_stats, gc_enable_stats, 0),
+    RB_BUILTIN_FUNCTION(14, gc_disable_stats, gc_disable_stats, 0),
+    RB_BUILTIN_FUNCTION(15, gc_stats_enabled, gc_stats_enabled, 0),
+    RB_BUILTIN_FUNCTION(16, gc_time, gc_time, 0),
+    RB_BUILTIN_FUNCTION(17, gc_heap_slots, gc_heap_slots, 0),
+    RB_BUILTIN_FUNCTION(18, gc_heap_slots_live_after_last_gc, gc_heap_slots_live_after_last_gc, 0),
+    RB_BUILTIN_FUNCTION(19, gc_total_mallocs, gc_total_mallocs, 0),
+    RB_BUILTIN_FUNCTION(20, gc_total_malloced_bytes, gc_total_malloced_bytes, 0),
     RB_BUILTIN_FUNCTION(-1, NULL, NULL, 0),
   };
 
@@ -73,6 +81,14 @@ COMPILER_WARNING_ERROR(-Wincompatible-po
   if (0) rb_builtin_function_check_arity0(builtin_inline_class_297);
   if (0) rb_builtin_function_check_arity0(builtin_inline_class_309);
   if (0) rb_builtin_function_check_arity0(builtin_inline_class_319);
+  if (0) rb_builtin_function_check_arity0(gc_enable_stats);
+  if (0) rb_builtin_function_check_arity0(gc_disable_stats);
+  if (0) rb_builtin_function_check_arity0(gc_stats_enabled);
+  if (0) rb_builtin_function_check_arity0(gc_time);
+  if (0) rb_builtin_function_check_arity0(gc_heap_slots);
+  if (0) rb_builtin_function_check_arity0(gc_heap_slots_live_after_last_gc);
+  if (0) rb_builtin_function_check_arity0(gc_total_mallocs);
+  if (0) rb_builtin_function_check_arity0(gc_total_malloced_bytes);
 COMPILER_WARNING_POP
 
   // load
diff -Nuarp ruby-3.3.11.a/include/ruby/internal/gc.h ruby-3.3.11.b/include/ruby/internal/gc.h
--- ruby-3.3.11.a/include/ruby/internal/gc.h	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/include/ruby/internal/gc.h	2026-03-29 01:56:52.897729156 -0400
@@ -431,6 +431,15 @@ void rb_gc_unregister_address(VALUE *val
  */
 void rb_gc_register_mark_object(VALUE object);
 
+VALUE rb_gc_enable_stats(void);
+VALUE rb_gc_disable_stats(void);
+VALUE rb_gc_stats_enabled(void);
+double rb_gc_total_time(void);
+VALUE rb_gc_heap_slots(void);
+VALUE rb_gc_heap_slots_live_after_last_gc(void);
+size_t rb_gc_total_mallocs(void);
+size_t rb_gc_total_malloced_bytes(void);
+
 RBIMPL_SYMBOL_EXPORT_END()
 
 /**
diff -Nuarp ruby-3.3.11.a/lib/rdoc/ri/paths.rb ruby-3.3.11.b/lib/rdoc/ri/paths.rb
--- ruby-3.3.11.a/lib/rdoc/ri/paths.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/lib/rdoc/ri/paths.rb	2026-03-29 01:56:10.347223801 -0400
@@ -10,7 +10,7 @@ module RDoc::RI::Paths
   #:stopdoc:
   require 'rbconfig'
 
-  version = RbConfig::CONFIG['ruby_version']
+  version = RbConfig::CONFIG['ruby_version_dir_name'] || RbConfig::CONFIG['ruby_version']
 
   BASE    = File.join RbConfig::CONFIG['ridir'], version
 
diff -Nuarp ruby-3.3.11.a/lib/rubygems/defaults.rb ruby-3.3.11.b/lib/rubygems/defaults.rb
--- ruby-3.3.11.a/lib/rubygems/defaults.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/lib/rubygems/defaults.rb	2026-03-29 01:56:10.347475825 -0400
@@ -35,7 +35,7 @@ module Gem
   # specified in the environment
 
   def self.default_dir
-    @default_dir ||= File.join(RbConfig::CONFIG["rubylibprefix"], "gems", RbConfig::CONFIG["ruby_version"])
+    @default_dir ||= File.join(RbConfig::CONFIG["rubylibprefix"], "gems", RbConfig::CONFIG["ruby_version_dir_name"] || RbConfig::CONFIG["ruby_version"])
   end
 
   ##
@@ -104,7 +104,8 @@ module Gem
     gem_dir = File.join(Gem.user_home, ".gem")
     gem_dir = File.join(Gem.data_home, "gem") unless File.exist?(gem_dir)
     parts = [gem_dir, ruby_engine]
-    parts << RbConfig::CONFIG["ruby_version"] unless RbConfig::CONFIG["ruby_version"].empty?
+    ruby_version_dir_name = RbConfig::CONFIG["ruby_version_dir_name"] || RbConfig::CONFIG["ruby_version"]
+    parts << ruby_version_dir_name unless ruby_version_dir_name.empty?
     File.join parts
   end
 
@@ -265,7 +266,7 @@ module Gem
     return nil unless RbConfig::CONFIG.key? "vendordir"
 
     File.join RbConfig::CONFIG["vendordir"], "gems",
-              RbConfig::CONFIG["ruby_version"]
+              RbConfig::CONFIG["ruby_version_dir_name"] || RbConfig::CONFIG["ruby_version"]
   end
 
   ##
diff -Nuarp ruby-3.3.11.a/loadpath.c ruby-3.3.11.b/loadpath.c
--- ruby-3.3.11.a/loadpath.c	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/loadpath.c	2026-03-29 01:56:30.355825212 -0400
@@ -65,22 +65,38 @@ const char ruby_initial_load_paths[] =
     RUBY_SEARCH_PATH "\0"
 #endif
 #ifndef NO_RUBY_SITE_LIB
+#ifdef RUBY_LIB_VERSION_BLANK
+    RUBY_SITE_LIB "\0"
+#else
     RUBY_SITE_LIB2 "\0"
+#endif
 #ifdef RUBY_THINARCH
     RUBY_SITE_ARCH_LIB_FOR(RUBY_THINARCH) "\0"
 #endif
     RUBY_SITE_ARCH_LIB_FOR(RUBY_SITEARCH) "\0"
+#ifndef RUBY_LIB_VERSION_BLANK
     RUBY_SITE_LIB "\0"
 #endif
+#endif
 
 #ifndef NO_RUBY_VENDOR_LIB
+#ifdef RUBY_LIB_VERSION_BLANK
+    RUBY_VENDOR_LIB "\0"
+#else
     RUBY_VENDOR_LIB2 "\0"
+#endif
 #ifdef RUBY_THINARCH
     RUBY_VENDOR_ARCH_LIB_FOR(RUBY_THINARCH) "\0"
 #endif
     RUBY_VENDOR_ARCH_LIB_FOR(RUBY_SITEARCH) "\0"
+#ifndef RUBY_LIB_VERSION_BLANK
     RUBY_VENDOR_LIB "\0"
 #endif
+#endif
+
+#ifdef RUBYGEMS_DIR
+    RUBYGEMS_DIR "\0"
+#endif
 
     RUBY_LIB "\0"
 #ifdef RUBY_THINARCH
diff -Nuarp ruby-3.3.11.a/miniprelude.c ruby-3.3.11.b/miniprelude.c
--- ruby-3.3.11.a/miniprelude.c	2026-03-25 20:05:06.000000000 -0400
+++ ruby-3.3.11.b/miniprelude.c	2026-03-29 01:56:57.388897621 -0400
@@ -1,7 +1,7 @@
 /* -*-c-*-
  THIS FILE WAS AUTOGENERATED BY template/prelude.c.tmpl. DO NOT EDIT.
 
- sources: ./ast, ./dir, ./gc, ./numeric, ./io, ./marshal, ./rjit, ./rjit_c, ./pack, ./trace_point, ./warning, ./array, ./kernel, ./ractor, ./symbol, ./timev, ./thread_sync, ./nilclass, ./prelude, ./gem_prelude, ./yjit
+ sources: ./tool/ruby_vm/helpers/c_escape, ./ast, ./dir, ./gc, ./numeric, ./io, ./marshal, ./rjit, ./rjit_c, ./pack, ./trace_point, ./warning, ./array, ./kernel, ./ractor, ./symbol, ./timev, ./thread_sync, ./nilclass, ./prelude, ./gem_prelude, ./yjit
 */
 #include "internal.h"
 #include "internal/ruby_parser.h"
@@ -16,14 +16,189 @@ COMPILER_WARNING_PUSH
 COMPILER_WARNING_IGNORED(-Wstring-concatenation)
 #endif
 
-static const char prelude_name0[] = "<internal:ast>";
+static const char prelude_name0[] = "<internal:tool/ruby_vm/helpers/c_escape>";
+static const struct {
+    RBIMPL_ATTR_NONSTRING() char L0[508]; /* 1..38 */
+    RBIMPL_ATTR_NONSTRING() char L38[505]; /* 39..61 */
+    RBIMPL_ATTR_NONSTRING() char L61[444]; /* 62..67 */
+    RBIMPL_ATTR_NONSTRING() char L67[444]; /* 68..73 */
+    RBIMPL_ATTR_NONSTRING() char L73[444]; /* 74..79 */
+    RBIMPL_ATTR_NONSTRING() char L79[444]; /* 80..85 */
+    RBIMPL_ATTR_NONSTRING() char L85[444]; /* 86..91 */
+    RBIMPL_ATTR_NONSTRING() char L91[444]; /* 92..97 */
+    RBIMPL_ATTR_NONSTRING() char L97[444]; /* 98..103 */
+    RBIMPL_ATTR_NONSTRING() char L103[444]; /* 104..109 */
+    RBIMPL_ATTR_NONSTRING() char L109[444]; /* 110..115 */
+    RBIMPL_ATTR_NONSTRING() char L115[506]; /* 116..128 */
+    RBIMPL_ATTR_NONSTRING() char L128[18]; /* 129..132 */
+} prelude_code0 = {
+#line 1 "tool/ruby_vm/helpers/c_escape.rb"
+""/* ! /your/favourite/path/to/ruby */
+""/* -*- Ruby -*- */
+""/* -*- frozen_string_literal: true; -*- */
+""/* -*- warn_indent: true; -*- */
+""/*  */
+""/* Copyright (c) 2017 Urabe, Shyouhei.  All rights reserved. */
+""/*  */
+""/* This file is  a part of the programming language  Ruby.  Permission is hereby */
+""/* granted, to  either redistribute and/or  modify this file, provided  that the */
+""/* conditions  mentioned in  the file  COPYING are  met.  Consult  the file  for */
+""/* details. */
+""
+"require 'securerandom'\n"
+"\n"
+"module RubyVM::CEscape\n"
+"  module_function\n"
+"\n"
+"\n"/* generate comment, with escaps. */
+"  def commentify str\n"
+"    unless str = str.dump[/\\A\"\\K.*(?=\"\\z)/]\n"
+"      raise Encoding::CompatibilityError, \"must be ASCII-compatible (#{str.encoding})\"\n"
+"    end\n"
+"    return \"/* #{str.gsub('*/', '*\\\\/').gsub('/*', '/\\\\*')} */\"\n"
+"  end\n"
+"\n"
+"\n"/* Mimic gensym of CL. */
+"  def gensym prefix = 'gensym_'\n"
+"    return as_tr_cpp \"#{prefix}#{SecureRandom.uuid}\"\n"
+"  end\n"
+"\n"
+"\n"/* Mimic AS_TR_CPP() of autoconf. */
+"  def as_tr_cpp name\n"
+"    q = name.b\n"
+"    q.gsub! %r/[^a-zA-Z0-9_]/m, '_'\n"
+"    q.gsub! %r/_+/, '_'\n"
+"    return q\n"
+"  end\n"
+"\n"
+,
+#line 39 "tool/ruby_vm/helpers/c_escape.rb"
+"\n"/* Section 6.10.4 of  ISO/IEC 9899:1999 specifies that the file  name used for */
+"\n"/* #line directive shall be a \"character string literal\".  So this is needed. */
+"\n"/*  */
+"\n"/* I'm  not sure  how  many  chars are  allowed  here,  though.  The  standard */
+"\n"/* specifies 4095 chars at most, after string concatenation (section 5.2.4.1). */
+"\n"/* But it is easy to have a path that is longer than that. */
+"\n"/*  */
+"\n"/* Here  we ignore  the standard.  Just create  single string  literal of  any */
+"\n"/* needed length. */
+"  def rstring2cstr str\n"
+"\n"/* I believe this is the fastest implementation done in pure-ruby. */
+"\n"/* Constants cached, gsub skips block evaluation, string literal optimized. */
+"    buf = str.b\n"
+"    buf.gsub! %r/./nm, RString2CStr\n"
+"    return %'\"#{buf}\"'\n"
+"  end\n"
+"\n"
+"  RString2CStr = {\n"
+"      \"\\x00\"=>  \"\\\\0\", \"\\x01\"=> \"\\\\x1\", \"\\x02\"=> \"\\\\x2\", \"\\x03\"=> \"\\\\x3\",\n"
+"      \"\\x04\"=> \"\\\\x4\", \"\\x05\"=> \"\\\\x5\", \"\\x06\"=> \"\\\\x6\",   \"\\a\"=>  \"\\\\a\",\n"
+"        \"\\b\"=>  \"\\\\b\",   \"\\t\"=>  \"\\\\t\",   \"\\n\"=>  \"\\\\n\",   \"\\v\"=>  \"\\\\v\",\n"
+"        \"\\f\"=>  \"\\\\f\",   \"\\r\"=>  \"\\\\r\", \"\\x0E\"=> \"\\\\xe\", \"\\x0F\"=> \"\\\\xf\",\n"
+"      \"\\x10\"=>\"\\\\x10\", \"\\x11\"=>\"\\\\x11\", \"\\x12\"=>\"\\\\x12\", \"\\x13\"=>\"\\\\x13\",\n"
+,
+#line 62 "tool/ruby_vm/helpers/c_escape.rb"
+"      \"\\x14\"=>\"\\\\x14\", \"\\x15\"=>\"\\\\x15\", \"\\x16\"=>\"\\\\x16\", \"\\x17\"=>\"\\\\x17\",\n"
+"      \"\\x18\"=>\"\\\\x18\", \"\\x19\"=>\"\\\\x19\", \"\\x1A\"=>\"\\\\x1a\",   \"\\e\"=>\"\\\\x1b\",\n"
+"      \"\\x1C\"=>\"\\\\x1c\", \"\\x1D\"=>\"\\\\x1d\", \"\\x1E\"=>\"\\\\x1e\", \"\\x1F\"=>\"\\\\x1f\",\n"
+"         \" \"=>    \" \",    \"!\"=>    \"!\",   \"\\\"\"=> \"\\\\\\\"\",    \"#\"=>    \"#\",\n"
+"         \"$\"=>    \"$\",    \"%\"=>    \"%\",    \"&\"=>    \"&\",    \"'\"=>    \"'\",\n"
+"         \"(\"=>    \"(\",    \")\"=>    \")\",    \"*\"=>    \"*\",    \"+\"=>    \"+\",\n"
+,
+#line 68 "tool/ruby_vm/helpers/c_escape.rb"
+"         \",\"=>    \",\",    \"-\"=>    \"-\",    \".\"=>    \".\",    \"/\"=>    \"/\",\n"
+"         \"0\"=>    \"0\",    \"1\"=>    \"1\",    \"2\"=>    \"2\",    \"3\"=>    \"3\",\n"
+"         \"4\"=>    \"4\",    \"5\"=>    \"5\",    \"6\"=>    \"6\",    \"7\"=>    \"7\",\n"
+"         \"8\"=>    \"8\",    \"9\"=>    \"9\",    \":\"=>    \":\",    \";\"=>    \";\",\n"
+"         \"<\"=>    \"<\",    \"=\"=>    \"=\",    \">\"=>    \">\",    \"?\"=>    \"?\",\n"
+"         \"@\"=>    \"@\",    \"A\"=>    \"A\",    \"B\"=>    \"B\",    \"C\"=>    \"C\",\n"
+,
+#line 74 "tool/ruby_vm/helpers/c_escape.rb"
+"         \"D\"=>    \"D\",    \"E\"=>    \"E\",    \"F\"=>    \"F\",    \"G\"=>    \"G\",\n"
+"         \"H\"=>    \"H\",    \"I\"=>    \"I\",    \"J\"=>    \"J\",    \"K\"=>    \"K\",\n"
+"         \"L\"=>    \"L\",    \"M\"=>    \"M\",    \"N\"=>    \"N\",    \"O\"=>    \"O\",\n"
+"         \"P\"=>    \"P\",    \"Q\"=>    \"Q\",    \"R\"=>    \"R\",    \"S\"=>    \"S\",\n"
+"         \"T\"=>    \"T\",    \"U\"=>    \"U\",    \"V\"=>    \"V\",    \"W\"=>    \"W\",\n"
+"         \"X\"=>    \"X\",    \"Y\"=>    \"Y\",    \"Z\"=>    \"Z\",    \"[\"=>    \"[\",\n"
+,
+#line 80 "tool/ruby_vm/helpers/c_escape.rb"
+"        \"\\\\\"=> \"\\\\\\\\\",    \"]\"=>    \"]\",    \"^\"=>    \"^\",    \"_\"=>    \"_\",\n"
+"         \"`\"=>    \"`\",    \"a\"=>    \"a\",    \"b\"=>    \"b\",    \"c\"=>    \"c\",\n"
+"         \"d\"=>    \"d\",    \"e\"=>    \"e\",    \"f\"=>    \"f\",    \"g\"=>    \"g\",\n"
+"         \"h\"=>    \"h\",    \"i\"=>    \"i\",    \"j\"=>    \"j\",    \"k\"=>    \"k\",\n"
+"         \"l\"=>    \"l\",    \"m\"=>    \"m\",    \"n\"=>    \"n\",    \"o\"=>    \"o\",\n"
+"         \"p\"=>    \"p\",    \"q\"=>    \"q\",    \"r\"=>    \"r\",    \"s\"=>    \"s\",\n"
+,
+#line 86 "tool/ruby_vm/helpers/c_escape.rb"
+"         \"t\"=>    \"t\",    \"u\"=>    \"u\",    \"v\"=>    \"v\",    \"w\"=>    \"w\",\n"
+"         \"x\"=>    \"x\",    \"y\"=>    \"y\",    \"z\"=>    \"z\",    \"{\"=>    \"{\",\n"
+"         \"|\"=>    \"|\",    \"}\"=>    \"}\",    \"~\"=>    \"~\", \"\\x7F\"=>\"\\\\x7f\",\n"
+"      \"\\x80\"=>\"\\\\x80\", \"\\x81\"=>\"\\\\x81\", \"\\x82\"=>\"\\\\x82\", \"\\x83\"=>\"\\\\x83\",\n"
+"      \"\\x84\"=>\"\\\\x84\", \"\\x85\"=>\"\\\\x85\", \"\\x86\"=>\"\\\\x86\", \"\\x87\"=>\"\\\\x87\",\n"
+"      \"\\x88\"=>\"\\\\x88\", \"\\x89\"=>\"\\\\x89\", \"\\x8A\"=>\"\\\\x8a\", \"\\x8B\"=>\"\\\\x8b\",\n"
+,
+#line 92 "tool/ruby_vm/helpers/c_escape.rb"
+"      \"\\x8C\"=>\"\\\\x8c\", \"\\x8D\"=>\"\\\\x8d\", \"\\x8E\"=>\"\\\\x8e\", \"\\x8F\"=>\"\\\\x8f\",\n"
+"      \"\\x90\"=>\"\\\\x90\", \"\\x91\"=>\"\\\\x91\", \"\\x92\"=>\"\\\\x92\", \"\\x93\"=>\"\\\\x93\",\n"
+"      \"\\x94\"=>\"\\\\x94\", \"\\x95\"=>\"\\\\x95\", \"\\x96\"=>\"\\\\x96\", \"\\x97\"=>\"\\\\x97\",\n"
+"      \"\\x98\"=>\"\\\\x98\", \"\\x99\"=>\"\\\\x99\", \"\\x9A\"=>\"\\\\x9a\", \"\\x9B\"=>\"\\\\x9b\",\n"
+"      \"\\x9C\"=>\"\\\\x9c\", \"\\x9D\"=>\"\\\\x9d\", \"\\x9E\"=>\"\\\\x9e\", \"\\x9F\"=>\"\\\\x9f\",\n"
+"      \"\\xA0\"=>\"\\\\xa0\", \"\\xA1\"=>\"\\\\xa1\", \"\\xA2\"=>\"\\\\xa2\", \"\\xA3\"=>\"\\\\xa3\",\n"
+,
+#line 98 "tool/ruby_vm/helpers/c_escape.rb"
+"      \"\\xA4\"=>\"\\\\xa4\", \"\\xA5\"=>\"\\\\xa5\", \"\\xA6\"=>\"\\\\xa6\", \"\\xA7\"=>\"\\\\xa7\",\n"
+"      \"\\xA8\"=>\"\\\\xa8\", \"\\xA9\"=>\"\\\\xa9\", \"\\xAA\"=>\"\\\\xaa\", \"\\xAB\"=>\"\\\\xab\",\n"
+"      \"\\xAC\"=>\"\\\\xac\", \"\\xAD\"=>\"\\\\xad\", \"\\xAE\"=>\"\\\\xae\", \"\\xAF\"=>\"\\\\xaf\",\n"
+"      \"\\xB0\"=>\"\\\\xb0\", \"\\xB1\"=>\"\\\\xb1\", \"\\xB2\"=>\"\\\\xb2\", \"\\xB3\"=>\"\\\\xb3\",\n"
+"      \"\\xB4\"=>\"\\\\xb4\", \"\\xB5\"=>\"\\\\xb5\", \"\\xB6\"=>\"\\\\xb6\", \"\\xB7\"=>\"\\\\xb7\",\n"
+"      \"\\xB8\"=>\"\\\\xb8\", \"\\xB9\"=>\"\\\\xb9\", \"\\xBA\"=>\"\\\\xba\", \"\\xBB\"=>\"\\\\xbb\",\n"
+,
+#line 104 "tool/ruby_vm/helpers/c_escape.rb"
+"      \"\\xBC\"=>\"\\\\xbc\", \"\\xBD\"=>\"\\\\xbd\", \"\\xBE\"=>\"\\\\xbe\", \"\\xBF\"=>\"\\\\xbf\",\n"
+"      \"\\xC0\"=>\"\\\\xc0\", \"\\xC1\"=>\"\\\\xc1\", \"\\xC2\"=>\"\\\\xc2\", \"\\xC3\"=>\"\\\\xc3\",\n"
+"      \"\\xC4\"=>\"\\\\xc4\", \"\\xC5\"=>\"\\\\xc5\", \"\\xC6\"=>\"\\\\xc6\", \"\\xC7\"=>\"\\\\xc7\",\n"
+"      \"\\xC8\"=>\"\\\\xc8\", \"\\xC9\"=>\"\\\\xc9\", \"\\xCA\"=>\"\\\\xca\", \"\\xCB\"=>\"\\\\xcb\",\n"
+"      \"\\xCC\"=>\"\\\\xcc\", \"\\xCD\"=>\"\\\\xcd\", \"\\xCE\"=>\"\\\\xce\", \"\\xCF\"=>\"\\\\xcf\",\n"
+"      \"\\xD0\"=>\"\\\\xd0\", \"\\xD1\"=>\"\\\\xd1\", \"\\xD2\"=>\"\\\\xd2\", \"\\xD3\"=>\"\\\\xd3\",\n"
+,
+#line 110 "tool/ruby_vm/helpers/c_escape.rb"
+"      \"\\xD4\"=>\"\\\\xd4\", \"\\xD5\"=>\"\\\\xd5\", \"\\xD6\"=>\"\\\\xd6\", \"\\xD7\"=>\"\\\\xd7\",\n"
+"      \"\\xD8\"=>\"\\\\xd8\", \"\\xD9\"=>\"\\\\xd9\", \"\\xDA\"=>\"\\\\xda\", \"\\xDB\"=>\"\\\\xdb\",\n"
+"      \"\\xDC\"=>\"\\\\xdc\", \"\\xDD\"=>\"\\\\xdd\", \"\\xDE\"=>\"\\\\xde\", \"\\xDF\"=>\"\\\\xdf\",\n"
+"      \"\\xE0\"=>\"\\\\xe0\", \"\\xE1\"=>\"\\\\xe1\", \"\\xE2\"=>\"\\\\xe2\", \"\\xE3\"=>\"\\\\xe3\",\n"
+"      \"\\xE4\"=>\"\\\\xe4\", \"\\xE5\"=>\"\\\\xe5\", \"\\xE6\"=>\"\\\\xe6\", \"\\xE7\"=>\"\\\\xe7\",\n"
+"      \"\\xE8\"=>\"\\\\xe8\", \"\\xE9\"=>\"\\\\xe9\", \"\\xEA\"=>\"\\\\xea\", \"\\xEB\"=>\"\\\\xeb\",\n"
+,
+#line 116 "tool/ruby_vm/helpers/c_escape.rb"
+"      \"\\xEC\"=>\"\\\\xec\", \"\\xED\"=>\"\\\\xed\", \"\\xEE\"=>\"\\\\xee\", \"\\xEF\"=>\"\\\\xef\",\n"
+"      \"\\xF0\"=>\"\\\\xf0\", \"\\xF1\"=>\"\\\\xf1\", \"\\xF2\"=>\"\\\\xf2\", \"\\xF3\"=>\"\\\\xf3\",\n"
+"      \"\\xF4\"=>\"\\\\xf4\", \"\\xF5\"=>\"\\\\xf5\", \"\\xF6\"=>\"\\\\xf6\", \"\\xF7\"=>\"\\\\xf7\",\n"
+"      \"\\xF8\"=>\"\\\\xf8\", \"\\xF9\"=>\"\\\\xf9\", \"\\xFA\"=>\"\\\\xfa\", \"\\xFB\"=>\"\\\\xfb\",\n"
+"      \"\\xFC\"=>\"\\\\xfc\", \"\\xFD\"=>\"\\\\xfd\", \"\\xFE\"=>\"\\\\xfe\", \"\\xFF\"=>\"\\\\xff\",\n"
+"  }.freeze\n"
+"  private_constant :RString2CStr\n"
+"end\n"
+"\n"
+"unless defined? ''.b\n"
+"  class String\n"
+"    def b\n"
+"      return dup.force_encoding 'binary'\n"
+,
+#line 129 "tool/ruby_vm/helpers/c_escape.rb"
+"    end\n"
+"  end\n"
+"end\n"
+#line 192 "miniprelude.c"
+};
+
+static const char prelude_name1[] = "<internal:ast>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[508]; /* 1..88 */
     RBIMPL_ATTR_NONSTRING() char L88[498]; /* 89..148 */
     RBIMPL_ATTR_NONSTRING() char L148[491]; /* 149..191 */
     RBIMPL_ATTR_NONSTRING() char L191[465]; /* 192..266 */
     RBIMPL_ATTR_NONSTRING() char L266[242]; /* 267..277 */
-} prelude_code0 = {
+} prelude_code1 = {
 #line 1 "ast.rb"
 ""/* for ast.c */
 ""
@@ -309,14 +484,14 @@ static const struct {
 "    end\n"
 "  end\n"
 "end\n"
-#line 313 "miniprelude.c"
+#line 488 "miniprelude.c"
 };
 
-static const char prelude_name1[] = "<internal:dir>";
+static const char prelude_name2[] = "<internal:dir>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[508]; /* 1..266 */
     RBIMPL_ATTR_NONSTRING() char L266[466]; /* 267..506 */
-} prelude_code1 = {
+} prelude_code2 = {
 #line 1 "dir.rb"
 ""/* An object of class \\Dir represents a directory in the underlying file system. */
 ""/*  */
@@ -336,12 +511,12 @@ static const struct {
 ""/* Some examples on this page use this simple file tree: */
 ""/*  */
 ""/*   example/ */
-""/*   \xE2\x94\x9C\xE2\x94\x80\xE2\x94\x80 config.h */
-""/*   \xE2\x94\x9C\xE2\x94\x80\xE2\x94\x80 lib/ */
-""/*   \xE2\x94\x82   \xE2\x94\x9C\xE2\x94\x80\xE2\x94\x80 song/ */
-""/*   \xE2\x94\x82   \xE2\x94\x82   \xE2\x94\x94\xE2\x94\x80\xE2\x94\x80 karaoke.rb */
-""/*   \xE2\x94\x82   \xE2\x94\x94\xE2\x94\x80\xE2\x94\x80 song.rb */
-""/*   \xE2\x94\x94\xE2\x94\x80\xE2\x94\x80 main.rb */
+""/*   \u251C\u2500\u2500 config.h */
+""/*   \u251C\u2500\u2500 lib/ */
+""/*   \u2502   \u251C\u2500\u2500 song/ */
+""/*   \u2502   \u2502   \u2514\u2500\u2500 karaoke.rb */
+""/*   \u2502   \u2514\u2500\u2500 song.rb */
+""/*   \u2514\u2500\u2500 main.rb */
 ""/*  */
 ""/* Others use the file tree for the */
 ""/* {Ruby project itself}[https://github.com/ruby/ruby]. */
@@ -825,17 +1000,18 @@ static const struct {
 "  end\n"
 "  alias fnmatch? fnmatch\n"
 "end if false\n"
-#line 829 "miniprelude.c"
+#line 1004 "miniprelude.c"
 };
 
-static const char prelude_name2[] = "<internal:gc>";
+static const char prelude_name3[] = "<internal:gc>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[498]; /* 1..76 */
     RBIMPL_ATTR_NONSTRING() char L76[507]; /* 77..253 */
     RBIMPL_ATTR_NONSTRING() char L253[506]; /* 254..299 */
-    RBIMPL_ATTR_NONSTRING() char L299[476]; /* 300..330 */
-    RBIMPL_ATTR_NONSTRING() char L330[39]; /* 331..333 */
-} prelude_code2 = {
+    RBIMPL_ATTR_NONSTRING() char L299[497]; /* 300..364 */
+    RBIMPL_ATTR_NONSTRING() char L364[469]; /* 365..413 */
+    RBIMPL_ATTR_NONSTRING() char L413[128]; /* 414..419 */
+} prelude_code3 = {
 #line 1 "gc.rb"
 ""/* for gc.c */
 ""
@@ -1165,29 +1341,117 @@ static const struct {
 "      ULL2NUM(rb_objspace.profile.marking_time_ns + rb_objspace.profile.sweeping_time_ns)\n"
 "    }\n"
 "  end\n"
+"\n"
+"\n"/* call-seq: */
+"\n"/*    GC.enable_stats\t  => true or false */
+"\n"/*  */
+"\n"/* Enables garbage collection statistics, returning <code>true</code> if garbage */
+"\n"/* collection statistics was already enabled. */
+"\n"/*  */
+"\n"/*    GC.enable_stats\t #=> false or true */
+"\n"/*    GC.enable_stats\t #=> true */
+"  def self.enable_stats\n"
+"    Primitive.gc_enable_stats\n"
+"  end\n"
+"\n"
+"\n"/*   call-seq: */
+"\n"/*    GC.disable_stats\t   => true or false */
+"\n"/*  */
+"\n"/* Disables garbage collection statistics, returning <code>true</code> if garbage */
+"\n"/* collection statistics was already disabled. */
+"\n"/*  */
+"\n"/*    GC.disable_stats\t  #=> false or true */
+"\n"/*    GC.disable_stats\t  #=> true */
+"  def self.disable_stats\n"
+"    Primitive.gc_disable_stats\n"
+"  end\n"
+"\n"
+"\n"/* call-seq: */
+"\n"/*    GC.stats_enabled?    => true or false */
+"\n"/*  */
+"\n"/* Check whether GC stats have been enabled. */
+"\n"/*  */
+"\n"/*    GC.stats_enabled?   #=> false or true */
+"  def self.stats_enabled?\n"
+"    Primitive.gc_stats_enabled\n"
+"  end\n"
+"\n"
+"\n"/*   call-seq: */
+"\n"/*    GC.time\t  => Integer */
+"\n"/*  */
+"\n"/* Returns the time spent during garbage collection while GC statistics collection */
+"\n"/* was enabled (in micro seconds). */
+"\n"/*  */
+"\n"/*    GC.time\t  #=> 20000 */
+,
+#line 365 "gc.rb"
+"  def self.time\n"
+"    Primitive.gc_time\n"
+"  end\n"
+"\n"
+"\n"/*  call-seq: */
+"\n"/* GC.heap_slots\t=> Integer */
+"\n"/*  */
+"\n"/*  Returns the number of heap slots available for object allocations. */
+"\n"/*  */
+"\n"/* GC.heap_slots\t#=> 10000 */
+"  def self.heap_slots\n"
+"    Primitive.gc_heap_slots\n"
+"  end\n"
+"\n"
+"\n"/* call-seq: */
+"\n"/*    GC.heap_slots_live_after_last_gc\t   => Integer */
+"\n"/*  */
+"\n"/* Returns the number of heap slots which were live after the last garbage collection. */
+"\n"/*  */
+"\n"/*    GC.heap_slots_live_after_last_gc\t   #=> 231223 */
+"  def self.heap_slots_live_after_last_gc\n"
+"    Primitive.gc_heap_slots_live_after_last_gc\n"
+"  end\n"
+"\n"
+"\n"/*   call-seq: */
+"\n"/*  GC.total_mallocs\t   => Integer */
+"\n"/*  */
+"\n"/*   Returns the number malloc calls. Might wrap around. */
+"\n"/*  */
+"\n"/*  GC.total_mallocs\t   #=> 324234323246 */
+"  def self.total_mallocs\n"
+"    Primitive.gc_total_mallocs\n"
+"  end\n"
+"\n"
+"\n"/*   call-seq: */
+"\n"/*  GC.total_malloced_bytes\t   => Integer */
+"\n"/*  */
+"\n"/*   Returns the number of bytes allocated. Might wrap around. */
+"\n"/*  */
+"\n"/*  GC.total_malloced_bytes\t   #=> 354656256432446 */
+"  def self.total_malloced_bytes\n"
+"    Primitive.gc_total_malloced_bytes\n"
+"  end\n"
+"\n"
 "end\n"
 "\n"
 "module ObjectSpace\n"
 "\n"/* Alias of GC.start */
 "  def garbage_collect full_mark: true, immediate_mark: true, immediate_sweep: true\n"
+,
+#line 414 "gc.rb"
 "    Primitive.gc_start_internal full_mark, immediate_mark, immediate_sweep, false\n"
 "  end\n"
 "\n"
-,
-#line 331 "gc.rb"
 "  module_function :garbage_collect\n"
 "end\n"
-#line 1181 "miniprelude.c"
+#line 1445 "miniprelude.c"
 };
 
-static const char prelude_name3[] = "<internal:numeric>";
+static const char prelude_name4[] = "<internal:numeric>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[507]; /* 1..101 */
     RBIMPL_ATTR_NONSTRING() char L101[508]; /* 102..200 */
     RBIMPL_ATTR_NONSTRING() char L200[503]; /* 201..284 */
     RBIMPL_ATTR_NONSTRING() char L284[507]; /* 285..357 */
     RBIMPL_ATTR_NONSTRING() char L357[216]; /* 358..373 */
-} prelude_code3 = {
+} prelude_code4 = {
 #line 1 "numeric.rb"
 "class Numeric\n"
 "\n"
@@ -1569,13 +1833,13 @@ static const struct {
 "  end\n"
 "\n"
 "end\n"
-#line 1573 "miniprelude.c"
+#line 1837 "miniprelude.c"
 };
 
-static const char prelude_name4[] = "<internal:io>";
+static const char prelude_name5[] = "<internal:io>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[449]; /* 1..137 */
-} prelude_code4 = {
+} prelude_code5 = {
 #line 1 "io.rb"
 "class IO\n"
 "\n"/* other IO methods are defined in io.c */
@@ -1713,13 +1977,13 @@ static const struct {
 "    Primitive.io_readline(sep, limit, chomp)\n"
 "  end\n"
 "end\n"
-#line 1717 "miniprelude.c"
+#line 1981 "miniprelude.c"
 };
 
-static const char prelude_name5[] = "<internal:marshal>";
+static const char prelude_name6[] = "<internal:marshal>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[202]; /* 1..41 */
-} prelude_code5 = {
+} prelude_code6 = {
 #line 1 "marshal.rb"
 "module Marshal\n"
 "\n"/* call-seq: */
@@ -1761,14 +2025,14 @@ static const struct {
 "    alias restore load\n"
 "  end\n"
 "end\n"
-#line 1765 "miniprelude.c"
+#line 2029 "miniprelude.c"
 };
 
-static const char prelude_name6[] = "<internal:rjit>";
+static const char prelude_name7[] = "<internal:rjit>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[498]; /* 1..31 */
     RBIMPL_ATTR_NONSTRING() char L31[197]; /* 32..42 */
-} prelude_code6 = {
+} prelude_code7 = {
 #line 1 "rjit.rb"
 "module RubyVM::RJIT\n"
 "\n"/* Return true if \\RJIT is enabled. */
@@ -1813,10 +2077,10 @@ static const struct {
 "  require 'ruby_vm/rjit/hooks'\n"
 "  require 'ruby_vm/rjit/stats'\n"
 "end\n"
-#line 1817 "miniprelude.c"
+#line 2081 "miniprelude.c"
 };
 
-static const char prelude_name7[] = "<internal:rjit_c>";
+static const char prelude_name8[] = "<internal:rjit_c>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[470]; /* 1..24 */
     RBIMPL_ATTR_NONSTRING() char L24[506]; /* 25..45 */
@@ -2020,7 +2284,7 @@ static const struct {
     RBIMPL_ATTR_NONSTRING() char L1629[505]; /* 1630..1659 */
     RBIMPL_ATTR_NONSTRING() char L1659[480]; /* 1660..1686 */
     RBIMPL_ATTR_NONSTRING() char L1686[146]; /* 1687..1696 */
-} prelude_code7 = {
+} prelude_code8 = {
 #line 1 "rjit_c.rb"
 ""/* frozen_string_literal: true */
 ""/* Part of this file is generated by tool/rjit/bindgen.rb. */
@@ -4119,13 +4383,13 @@ static const struct {
 "\n"
 "\n"/* # RJIT bindgen end ### */
 "end if Primitive.rjit_enabled_p\n"
-#line 4123 "miniprelude.c"
+#line 4387 "miniprelude.c"
 };
 
-static const char prelude_name8[] = "<internal:pack>";
+static const char prelude_name9[] = "<internal:pack>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[273]; /* 1..32 */
-} prelude_code8 = {
+} prelude_code9 = {
 #line 1 "pack.rb"
 "class Array\n"
 "\n"/*  call-seq: */
@@ -4158,16 +4422,16 @@ static const struct {
 "    Primitive.pack_unpack1(fmt, offset)\n"
 "  end\n"
 "end\n"
-#line 4162 "miniprelude.c"
+#line 4426 "miniprelude.c"
 };
 
-static const char prelude_name9[] = "<internal:trace_point>";
+static const char prelude_name10[] = "<internal:trace_point>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[508]; /* 1..237 */
     RBIMPL_ATTR_NONSTRING() char L237[506]; /* 238..327 */
     RBIMPL_ATTR_NONSTRING() char L327[495]; /* 328..401 */
     RBIMPL_ATTR_NONSTRING() char L401[215]; /* 402..419 */
-} prelude_code9 = {
+} prelude_code10 = {
 #line 1 "trace_point.rb"
 ""/* loaded from vm_trace.c */
 ""
@@ -4593,13 +4857,13 @@ static const struct {
 "    Primitive.tracepoint_attr_instruction_sequence\n"
 "  end\n"
 "end\n"
-#line 4597 "miniprelude.c"
+#line 4861 "miniprelude.c"
 };
 
-static const char prelude_name10[] = "<internal:warning>";
+static const char prelude_name11[] = "<internal:warning>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[182]; /* 1..54 */
-} prelude_code10 = {
+} prelude_code11 = {
 #line 1 "warning.rb"
 ""/* encoding: utf-8 */
 ""/* frozen-string-literal: true */
@@ -4654,15 +4918,15 @@ static const struct {
 "    Primitive.rb_warn_m(msgs, uplevel, category)\n"
 "  end\n"
 "end\n"
-#line 4658 "miniprelude.c"
+#line 4922 "miniprelude.c"
 };
 
-static const char prelude_name11[] = "<internal:array>";
+static const char prelude_name12[] = "<internal:array>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[493]; /* 1..102 */
     RBIMPL_ATTR_NONSTRING() char L102[478]; /* 103..150 */
     RBIMPL_ATTR_NONSTRING() char L150[175]; /* 151..158 */
-} prelude_code11 = {
+} prelude_code12 = {
 #line 1 "array.rb"
 "class Array\n"
 "\n"/* call-seq: */
@@ -4825,15 +5089,15 @@ static const struct {
 "    end\n"
 "  end\n"
 "end\n"
-#line 4829 "miniprelude.c"
+#line 5093 "miniprelude.c"
 };
 
-static const char prelude_name12[] = "<internal:kernel>";
+static const char prelude_name13[] = "<internal:kernel>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[446]; /* 1..130 */
     RBIMPL_ATTR_NONSTRING() char L130[508]; /* 131..192 */
     RBIMPL_ATTR_NONSTRING() char L192[464]; /* 193..313 */
-} prelude_code12 = {
+} prelude_code13 = {
 #line 1 "kernel.rb"
 "module Kernel\n"
 "\n"/*  */
@@ -5151,10 +5415,10 @@ static const struct {
 "    end\n"
 "  end\n"
 "end\n"
-#line 5155 "miniprelude.c"
+#line 5419 "miniprelude.c"
 };
 
-static const char prelude_name13[] = "<internal:ractor>";
+static const char prelude_name14[] = "<internal:ractor>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[502]; /* 1..281 */
     RBIMPL_ATTR_NONSTRING() char L281[503]; /* 282..360 */
@@ -5164,7 +5428,7 @@ static const struct {
     RBIMPL_ATTR_NONSTRING() char L719[485]; /* 720..767 */
     RBIMPL_ATTR_NONSTRING() char L767[487]; /* 768..838 */
     RBIMPL_ATTR_NONSTRING() char L838[230]; /* 839..854 */
-} prelude_code13 = {
+} prelude_code14 = {
 #line 1 "ractor.rb"
 ""/* \\Ractor is an Actor-model abstraction for Ruby that provides thread-safe parallel execution. */
 ""/*  */
@@ -6033,13 +6297,13 @@ static const struct {
 "    }\n"
 "  end\n"
 "end\n"
-#line 6037 "miniprelude.c"
+#line 6301 "miniprelude.c"
 };
 
-static const char prelude_name14[] = "<internal:symbol>";
+static const char prelude_name15[] = "<internal:symbol>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[74]; /* 1..14 */
-} prelude_code14 = {
+} prelude_code15 = {
 #line 1 "symbol.rb"
 "class Symbol\n"
 "\n"/* call-seq: */
@@ -6054,15 +6318,15 @@ static const struct {
 "\n"
 "  alias intern to_sym\n"
 "end\n"
-#line 6058 "miniprelude.c"
+#line 6322 "miniprelude.c"
 };
 
-static const char prelude_name15[] = "<internal:timev>";
+static const char prelude_name16[] = "<internal:timev>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[477]; /* 1..394 */
     RBIMPL_ATTR_NONSTRING() char L394[507]; /* 395..412 */
     RBIMPL_ATTR_NONSTRING() char L412[78]; /* 413..416 */
-} prelude_code15 = {
+} prelude_code16 = {
 #line 1 "timev.rb"
 ""/* A +Time+ object represents a date and time: */
 ""/*  */
@@ -6483,14 +6747,14 @@ static const struct {
 "    Primitive.time_init_args(year, mon, mday, hour, min, sec, zone)\n"
 "  end\n"
 "end\n"
-#line 6487 "miniprelude.c"
+#line 6751 "miniprelude.c"
 };
 
-static const char prelude_name16[] = "<internal:thread_sync>";
+static const char prelude_name17[] = "<internal:thread_sync>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[500]; /* 1..38 */
     RBIMPL_ATTR_NONSTRING() char L38[446]; /* 39..69 */
-} prelude_code16 = {
+} prelude_code17 = {
 #line 1 "thread_sync.rb"
 "class Thread\n"
 "  class Queue\n"
@@ -6562,13 +6826,13 @@ static const struct {
 "    alias_method :<<, :push\n"
 "  end\n"
 "end\n"
-#line 6566 "miniprelude.c"
+#line 6830 "miniprelude.c"
 };
 
-static const char prelude_name17[] = "<internal:nilclass>";
+static const char prelude_name18[] = "<internal:nilclass>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[98]; /* 1..26 */
-} prelude_code17 = {
+} prelude_code18 = {
 #line 1 "nilclass.rb"
 "class NilClass\n"
 "\n"/*  */
@@ -6595,13 +6859,13 @@ static const struct {
 "    return 0.0\n"
 "  end\n"
 "end\n"
-#line 6599 "miniprelude.c"
+#line 6863 "miniprelude.c"
 };
 
-static const char prelude_name18[] = "<internal:prelude>";
+static const char prelude_name19[] = "<internal:prelude>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[352]; /* 1..32 */
-} prelude_code18 = {
+} prelude_code19 = {
 #line 1 "prelude.rb"
 "class Binding\n"
 "\n"/* :nodoc: */
@@ -6634,14 +6898,14 @@ static const struct {
 "    klass.new(self, *args, &block)\n"
 "  end unless instance_methods.include?(:to_set)\n"/* RJIT could already load this from builtin prelude */
 "end\n"
-#line 6638 "miniprelude.c"
+#line 6902 "miniprelude.c"
 };
 
-static const char prelude_name19[] = "<internal:gem_prelude>";
+static const char prelude_name20[] = "<internal:gem_prelude>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[478]; /* 1..25 */
     RBIMPL_ATTR_NONSTRING() char L25[74]; /* 26..29 */
-} prelude_code19 = {
+} prelude_code20 = {
 #line 1 "gem_prelude.rb"
 "begin\n"
 "  require 'rubygems'\n"
@@ -6673,10 +6937,10 @@ static const struct {
 "  warn \"`syntax_suggest' was not loaded.\"\n"
 "end if defined?(SyntaxSuggest)\n"
 "\n"
-#line 6677 "miniprelude.c"
+#line 6941 "miniprelude.c"
 };
 
-static const char prelude_name20[] = "<internal:yjit>";
+static const char prelude_name21[] = "<internal:yjit>";
 static const struct {
     RBIMPL_ATTR_NONSTRING() char L0[506]; /* 1..39 */
     RBIMPL_ATTR_NONSTRING() char L39[506]; /* 40..60 */
@@ -6713,7 +6977,7 @@ static const struct {
     RBIMPL_ATTR_NONSTRING() char L454[461]; /* 455..469 */
     RBIMPL_ATTR_NONSTRING() char L469[472]; /* 470..487 */
     RBIMPL_ATTR_NONSTRING() char L487[244]; /* 488..496 */
-} prelude_code20 = {
+} prelude_code21 = {
 #line 1 "yjit.rb"
 ""/* frozen_string_literal: true */
 ""/* :markup: markdown */
@@ -7278,7 +7542,7 @@ static const struct {
 "    end\n"
 "  end\n"
 "end\n"
-#line 7282 "miniprelude.c"
+#line 7546 "miniprelude.c"
 };
 
 COMPILER_WARNING_POP
@@ -7309,27 +7573,28 @@ rb_builtin_ast(const char *feature_name,
     size_t namelen = strlen(feature_name);
     rb_ast_t *ast = 0;
 
-    if ((ast = PRELUDE_AST(0, *name_str, 20)) != 0) return ast;
-    if ((ast = PRELUDE_AST(1, *name_str, 158)) != 0) return ast;
-    if ((ast = PRELUDE_AST(2, *name_str, 11)) != 0) return ast;
-    if ((ast = PRELUDE_AST(3, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(0, *name_str, 13)) != 0) return ast;
+    if ((ast = PRELUDE_AST(1, *name_str, 20)) != 0) return ast;
+    if ((ast = PRELUDE_AST(2, *name_str, 158)) != 0) return ast;
+    if ((ast = PRELUDE_AST(3, *name_str, 11)) != 0) return ast;
     if ((ast = PRELUDE_AST(4, *name_str, 1)) != 0) return ast;
     if ((ast = PRELUDE_AST(5, *name_str, 1)) != 0) return ast;
     if ((ast = PRELUDE_AST(6, *name_str, 1)) != 0) return ast;
-    if ((ast = PRELUDE_AST(7, *name_str, 4)) != 0) return ast;
-    if ((ast = PRELUDE_AST(8, *name_str, 1)) != 0) return ast;
-    if ((ast = PRELUDE_AST(9, *name_str, 50)) != 0) return ast;
-    if ((ast = PRELUDE_AST(10, *name_str, 4)) != 0) return ast;
-    if ((ast = PRELUDE_AST(11, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(7, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(8, *name_str, 4)) != 0) return ast;
+    if ((ast = PRELUDE_AST(9, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(10, *name_str, 50)) != 0) return ast;
+    if ((ast = PRELUDE_AST(11, *name_str, 4)) != 0) return ast;
     if ((ast = PRELUDE_AST(12, *name_str, 1)) != 0) return ast;
-    if ((ast = PRELUDE_AST(13, *name_str, 239)) != 0) return ast;
-    if ((ast = PRELUDE_AST(14, *name_str, 1)) != 0) return ast;
-    if ((ast = PRELUDE_AST(15, *name_str, 216)) != 0) return ast;
-    if ((ast = PRELUDE_AST(16, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(13, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(14, *name_str, 239)) != 0) return ast;
+    if ((ast = PRELUDE_AST(15, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(16, *name_str, 216)) != 0) return ast;
     if ((ast = PRELUDE_AST(17, *name_str, 1)) != 0) return ast;
     if ((ast = PRELUDE_AST(18, *name_str, 1)) != 0) return ast;
     if ((ast = PRELUDE_AST(19, *name_str, 1)) != 0) return ast;
-    if ((ast = PRELUDE_AST(20, *name_str, 10)) != 0) return ast;
+    if ((ast = PRELUDE_AST(20, *name_str, 1)) != 0) return ast;
+    if ((ast = PRELUDE_AST(21, *name_str, 10)) != 0) return ast;
     return ast;
 }
 
diff -Nuarp ruby-3.3.11.a/ruby.c ruby-3.3.11.b/ruby.c
--- ruby-3.3.11.a/ruby.c	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/ruby.c	2026-03-29 01:56:34.173079012 -0400
@@ -1724,10 +1724,14 @@ proc_options(long argc, char **argv, rub
 
 void Init_builtin_features(void);
 
+/* abrt.c */
+void Init_abrt(void);
+
 static void
 ruby_init_prelude(void)
 {
     Init_builtin_features();
+    Init_abrt();
     rb_const_remove(rb_cObject, rb_intern_const("TMP_RUBY_PREFIX"));
 }
 
diff -Nuarp ruby-3.3.11.a/spec/ruby/core/kernel/require_spec.rb ruby-3.3.11.b/spec/ruby/core/kernel/require_spec.rb
--- ruby-3.3.11.a/spec/ruby/core/kernel/require_spec.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/spec/ruby/core/kernel/require_spec.rb	2026-03-29 01:56:34.173237735 -0400
@@ -25,6 +25,8 @@ describe "Kernel#require" do
     out = ruby_exe("puts $LOADED_FEATURES", options: '--disable-gems --disable-did-you-mean')
     features = out.lines.map { |line| File.basename(line.chomp, '.*') }
 
+    # Ignore ABRT
+    features -= %w[abrt]
     # Ignore CRuby internals
     features -= %w[encdb transdb windows_1252]
     features.reject! { |feature| feature.end_with?('-fake') }
diff -Nuarp ruby-3.3.11.a/template/Makefile.in ruby-3.3.11.b/template/Makefile.in
--- ruby-3.3.11.a/template/Makefile.in	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/template/Makefile.in	2026-03-29 01:56:43.570744354 -0400
@@ -90,6 +90,7 @@ cflags = @cflags@
 optflags = @optflags@
 debugflags = @debugflags@
 warnflags = @warnflags@ @strict_warnflags@
+hardenflags = @hardenflags@
 cppflags = @cppflags@
 incflags = @incflags@
 RUBY_DEVEL = @RUBY_DEVEL@ # "yes" or empty
diff -Nuarp ruby-3.3.11.a/template/ruby.pc.in ruby-3.3.11.b/template/ruby.pc.in
--- ruby-3.3.11.a/template/ruby.pc.in	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/template/ruby.pc.in	2026-03-29 01:56:10.347005668 -0400
@@ -2,6 +2,7 @@ MAJOR=@MAJOR@
 MINOR=@MINOR@
 TEENY=@TEENY@
 ruby_version=@ruby_version@
+ruby_version_dir_name=@ruby_version_dir_name@
 RUBY_API_VERSION=@RUBY_API_VERSION@
 RUBY_PROGRAM_VERSION=@RUBY_PROGRAM_VERSION@
 arch=@arch@
diff -Nuarp ruby-3.3.11.a/template/verconf.h.tmpl ruby-3.3.11.b/template/verconf.h.tmpl
--- ruby-3.3.11.a/template/verconf.h.tmpl	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/template/verconf.h.tmpl	2026-03-29 01:56:30.356025686 -0400
@@ -36,6 +36,9 @@
 % if C["RUBY_SEARCH_PATH"]
 #define RUBY_SEARCH_PATH		"${RUBY_SEARCH_PATH}"
 % end
+% if C["rubygemsdir"]
+#define RUBYGEMS_DIR			"${rubygemsdir}"
+% end
 %
 % R = {}
 % R["ruby_version"] = '"RUBY_LIB_VERSION"'
diff -Nuarp ruby-3.3.11.a/test/rubygems/test_gem.rb ruby-3.3.11.b/test/rubygems/test_gem.rb
--- ruby-3.3.11.a/test/rubygems/test_gem.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/test/rubygems/test_gem.rb	2026-03-29 01:56:10.347606338 -0400
@@ -1365,7 +1365,8 @@ class TestGem < Gem::TestCase
 
   def test_self_user_dir
     parts = [@userhome, ".gem", Gem.ruby_engine]
-    parts << RbConfig::CONFIG["ruby_version"] unless RbConfig::CONFIG["ruby_version"].empty?
+    ruby_version_dir_name = RbConfig::CONFIG["ruby_version_dir_name"] || RbConfig::CONFIG["ruby_version"]
+    parts << ruby_version_dir_name unless ruby_version_dir_name.empty?
 
     FileUtils.mkdir_p File.join(parts)
 
@@ -1441,7 +1442,7 @@ class TestGem < Gem::TestCase
     vendordir(File.join(@tempdir, "vendor")) do
       expected =
         File.join RbConfig::CONFIG["vendordir"], "gems",
-                  RbConfig::CONFIG["ruby_version"]
+                  RbConfig::CONFIG["ruby_version_dir_name"] || RbConfig::CONFIG["ruby_version"]
 
       assert_equal expected, Gem.vendor_dir
     end
diff -Nuarp ruby-3.3.11.a/tool/m4/ruby_append_option.m4 ruby-3.3.11.b/tool/m4/ruby_append_option.m4
--- ruby-3.3.11.a/tool/m4/ruby_append_option.m4	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/tool/m4/ruby_append_option.m4	2026-03-29 01:56:43.570960598 -0400
@@ -3,3 +3,7 @@ AC_DEFUN([RUBY_APPEND_OPTION],
 	[# RUBY_APPEND_OPTION($1)
 	AS_CASE([" [$]{$1-} "],
 	[*" $2 "*], [], ['  '], [ $1="$2"], [ $1="[$]$1 $2"])])dnl
+AC_DEFUN([RUBY_PREPEND_OPTION],
+	[# RUBY_APPEND_OPTION($1)
+	AS_CASE([" [$]{$1-} "],
+	[*" $2 "*], [], ['  '], [ $1="$2"], [ $1="$2 [$]$1"])])dnl
diff -Nuarp ruby-3.3.11.a/tool/m4/ruby_try_cflags.m4 ruby-3.3.11.b/tool/m4/ruby_try_cflags.m4
--- ruby-3.3.11.a/tool/m4/ruby_try_cflags.m4	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/tool/m4/ruby_try_cflags.m4	2026-03-29 01:56:43.571163021 -0400
@@ -22,3 +22,20 @@ AC_DEFUN([RUBY_TRY_CFLAGS], [
 	AC_MSG_RESULT(no)],
 	[$4], [$5])
 ])dnl
+
+AC_DEFUN([_RUBY_TRY_CFLAGS_PREPEND], [
+    RUBY_WERROR_FLAG([
+    CFLAGS="$1 [$]CFLAGS"
+    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[$4]], [[$5]])],
+	[$2], [$3])
+    ])dnl
+])dnl
+AC_DEFUN([RUBY_TRY_CFLAGS_PREPEND], [
+    AC_MSG_CHECKING([whether ]$1[ is accepted as CFLAGS])dnl
+    _RUBY_TRY_CFLAGS_PREPEND([$1],
+	[$2
+	AC_MSG_RESULT(yes)],
+	[$3
+	AC_MSG_RESULT(no)],
+	[$4], [$5])
+])dnl
diff -Nuarp ruby-3.3.11.a/tool/mkconfig.rb ruby-3.3.11.b/tool/mkconfig.rb
--- ruby-3.3.11.a/tool/mkconfig.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/tool/mkconfig.rb	2026-03-29 01:56:15.587135101 -0400
@@ -115,7 +115,7 @@ File.foreach "config.status" do |line|
     val = val.gsub(/\$(?:\$|\{?(\w+)\}?)/) {$1 ? "$(#{$1})" : $&}.dump
     case name
     when /^prefix$/
-      val = "(TOPDIR || DESTDIR + #{val})"
+      val = "(((TOPDIR && TOPDIR.empty?) ? nil : TOPDIR) || DESTDIR + #{val})"
     when /^ARCH_FLAG$/
       val = "arch_flag || #{val}" if universal
     when /^UNIVERSAL_ARCHNAMES$/
diff -Nuarp ruby-3.3.11.a/tool/rbinstall.rb ruby-3.3.11.b/tool/rbinstall.rb
--- ruby-3.3.11.a/tool/rbinstall.rb	2026-03-25 20:05:04.000000000 -0400
+++ ruby-3.3.11.b/tool/rbinstall.rb	2026-03-29 01:56:30.356131317 -0400
@@ -359,6 +359,7 @@ if CONFIG["vendordir"]
   vendorlibdir = CONFIG["vendorlibdir"]
   vendorarchlibdir = CONFIG["vendorarchdir"]
 end
+rubygemsdir = CONFIG["rubygemsdir"]
 mandir = CONFIG["mandir", true]
 docdir = CONFIG["docdir", true]
 enable_shared = CONFIG["ENABLE_SHARED"] == 'yes'
@@ -453,7 +454,7 @@ end
 
 install?(:doc, :rdoc) do
   if $rdocdir
-    ridatadir = File.join(CONFIG['ridir'], CONFIG['ruby_version'], "system")
+    ridatadir = File.join(CONFIG['ridir'], CONFIG['ruby_version_dir_name'] || CONFIG['ruby_version'], "system")
     prepare "rdoc", ridatadir
     install_recursive($rdocdir, ridatadir, :no_install => rdoc_noinst, :mode => $data_mode)
   end
@@ -595,7 +596,16 @@ end
 install?(:local, :comm, :lib) do
   prepare "library scripts", rubylibdir
   noinst = %w[*.txt *.rdoc *.gemspec]
+  # Bundler carries "rubygems.rb" file, so it must be specialcased :/
+  noinst += %w[rubygems.rb rubygems/ bundler.rb bundler/] if rubygemsdir
   install_recursive(File.join(srcdir, "lib"), rubylibdir, :no_install => noinst, :mode => $data_mode)
+  if rubygemsdir
+    noinst = %w[*.txt *.rdoc *.gemspec]
+    install_recursive(File.join(srcdir, "lib", "rubygems"), File.join(rubygemsdir, "rubygems"), :no_install => noinst, :mode => $data_mode)
+    install(File.join(srcdir, "lib", "rubygems.rb"), File.join(rubygemsdir, "rubygems.rb"), :mode => $data_mode)
+    install_recursive(File.join(srcdir, "lib", "bundler"), File.join(rubylibdir, "bundler"), :no_install => noinst, :mode => $data_mode)
+    install(File.join(srcdir, "lib", "bundler.rb"), rubylibdir, :mode => $data_mode)
+  end
 end
 
 install?(:local, :comm, :hdr, :'comm-hdr') do
